lib/fluent/plugin/buffer/file_single_chunk.rb

```diff
@@ -0,0 +1,315 @@
+              # Retry is one possible approach but it may cause livelock under limited resources or high load environment.
+              # So we ignore such errors for now and log better message instead.
+              # "Too many open files" should be fixed by proper buffer configuration and system setting.
+              raise "can't enqueue buffer file and failed to restore. This may causes inconsistent state: path = #{@path}, error = '#{e}', retry error = '#{re}'"
+            else
+              raise "can't enqueue buffer file: path = #{@path}, error = '#{e}'"
+            end
+          end
+
+          @path = new_chunk_path
+
+          super
+        end
+
+        def close
+          super
+          size = @chunk.size
+          @chunk.close
+          if size == 0
+            File.unlink(@path)
+          end
+        end
+
+        def purge
+          super
+          @chunk.close
+          @bytesize = @size = @adding_bytes = @adding_size = 0
+          File.unlink(@path)
+        end
+
+        def read(**kwargs)
+          @chunk.seek(0, IO::SEEK_SET)
+          @chunk.read
+        end
+
+        def open(**kwargs, &block)
+          @chunk.seek(0, IO::SEEK_SET)
+          val = yield @chunk
+          @chunk.seek(0, IO::SEEK_END) if self.staged?
+          val
+        end
+
+        def self.assume_chunk_state(path)
+          return :unknown unless path.end_with?(PATH_SUFFIX)
+
+          if PATH_REGEXP =~ path
+            $1 == 'b' ? :staged : :queued
+          else
+            # files which matches to glob of buffer file pattern
+            # it includes files which are created by out_file
+            :unknown
+          end
+        end
+
+        def self.unique_id_and_key_from_path(path)
+          base = File.basename(path)
+          res = PATH_REGEXP =~ base
+          return nil unless res
+
+          key = base[4..res - 1]  # remove 'sfb.' and '.'
+          hex_id = base[res + 2..-5] # remove '.' and '.bug' 
+          unique_id = hex_id.scan(/../).map {|x| x.to_i(16) }.pack('C*')
+          [unique_id, key]
+        end
+
+        def self.generate_stage_chunk_path(path, key, unique_id)
+          pos = path.index('.*.')
+          raise "BUG: buffer chunk path on stage MUST have '.*.'" unless pos
+
+          prefix = path[0...pos]
+          suffix = path[(pos + 3)..-1]
+
+          chunk_id = Fluent::UniqueId.hex(unique_id)
+          state = 'b'
+          "#{prefix}.#{key}.#{state}#{chunk_id}.#{suffix}"
+        end
+
+        def self.generate_queued_chunk_path(path, unique_id)
+          chunk_id = Fluent::UniqueId.hex(unique_id)
+          staged_path = ".b#{chunk_id}."
+          if path.index(staged_path)
+            path.sub(staged_path, ".q#{chunk_id}.")
+          else # for unexpected cases (ex: users rename files while opened by fluentd)
+            path + ".q#{chunk_id}.chunk"
+          end
+        end
+
+        def restore_metadata
```

Comment:
Doesn't file_single_chunk support file_chunk's metadata restore?(this is just confirmation)

Reply:
We can support it with metadata header in the future.
Just now, support only one field key or tag.